pymeasure/instruments/thorlabs/thorlabspm100usb.py

```diff
@@ -23,74 +23,97 @@
 #
 
 import logging
+
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import truncated_range
+
 log = logging.getLogger(__name__)
 log.addHandler(logging.NullHandler())
 
-from pymeasure.instruments import Instrument, RangeException
-
 
 class ThorlabsPM100USB(Instrument):
-    """Represents Thorlabs PM100USB powermeter"""
-
-    # TODO: refactor to check if the sensor wavelength is adjustable
-    wavelength = Instrument.control("SENSE:CORR:WAV?", "SENSE:CORR:WAV %g",
-                                    "Wavelength in nm; not set outside of range")
-    
-    # TODO: refactor to check if the sensor is a power sensor
-    power = Instrument.measurement("MEAS:POW?", "Power, in Watts")
-    
-    wavelength_min = Instrument.measurement("SENS:CORR:WAV? MIN", "Get minimum wavelength, in nm")
-    
-    wavelength_max = Instrument.measurement("SENS:CORR:WAV? MAX", "Get maximum wavelength, in nm")
-
+    """Represents Thorlabs PM100USB powermeter."""
 
     def __init__(self, adapter, **kwargs):
         super(ThorlabsPM100USB, self).__init__(
-            adapter, "ThorlabsPM100USB powermeter", **kwargs)
+            adapter, "ThorlabsPM100USB powermeter", **kwargs
+        )
         self.timout = 3000
-        self.sensor()
-
-    def measure_power(self, wavelength):
-        """Set wavelength in nm and get power in W
-        If wavelength is out of range it will be set to range limit"""
-        if wavelength < self.wavelength_min:
-            raise RangeException("Wavelength %.2f nm out of range: using minimum wavelength: %.2f nm" % (
-                wavelength, self.wavelength_min))
-            # explicit setting wavelenghth, althought it would be automatically set
-            wavelength = self.wavelength_min
-        if wavelength > self.wavelength_max:
-            raise RangeException("Wavelength %.2f nm out of range: using maximum wavelength: %.2f nm" % (
-                wavelength, self.wavelength_max))
-            wavelength = self.wavelength_max
-        self.wavelength = wavelength
-        return self.power
-
-    def sensor(self):
-        "Get sensor info"
-        response = self.ask("SYST:SENSOR:IDN?").split(',')
+        self._set_flags()
+
+    wavelength_min = Instrument.measurement(
+        "SENS:CORR:WAV? MIN", "Minimum wavelength, in nm"
+    )
+
+    wavelength_max = Instrument.measurement(
+        "SENS:CORR:WAV? MAX", "Maximum wavelength, in nm"
+    )
+
+    @property
+    def wavelength(self):
+        """Wavelength in nm."""
+        value = self.values("SENSE:CORR:WAV?")[0]
+        return value
+
+    @wavelength.setter
+    def wavelength(self, value):
+        """Wavelength in nm."""
+        if self.wavelength_settable:
+            value = truncated_range(value, [self.wavelength_min, self.wavelength_max])
```

Comment:
watch out here, this means every time you set the wavelength, you have 3 communications with the instrument, as wavelength_min and wavelength_max will be queried anew. If those are constant properties (e.g. instrument or sensor head parameters), it might be better to store the values in, e.g. `self._wavelength_min`, `self._wavelength_max` and fill those values once in the initializer of the class.

Reply:
I added a check to only request the minimum and maximum wavelength once: a76423a

I put it in the wavelength setter and not in the class init, since wavelength might not be settable by some instruments. This way, `AttributeError` is raised in this case without the need for an additional check in `__init__`.

Reply:
Nicely spotted, good point!